### ISSUE TITLE
React Lazy with TransitionGroup 

### ISSUE THREAD

**Issue by PodaruDragos** [reactjs/react-transition-group#580]
I have tried to animate views that are lazy loaded with `React.lazy();`

i have just picked the example at [link](http://reactcommunity.org/react-transition-group/with-react-router/) and modified the Components to be lazy loaded.

then i suspect when trying to unmount `React.Suspense` breaks something. [codesandbox](https://codesandbox.io/s/csstransition-react-router-dqqwq)

**Comment by Janaka-Steph** [comment]
Same here, can't find a proper solution. I found [this article](https://itnext.io/react-router-transitions-with-lazy-loading-2faa7a1d24a) interesting, even if not using react-transition-group. It says that "[lazy loading] breaks the initial route animation when you refresh the page, because the async operation uptakes the single-threaded run time in JavaScript."

**Comment by PodaruDragos** [comment]
@Janaka-Steph i made an example that "kinda" works.. maybe it will help you out

[code](https://codesandbox.io/s/delicate-butterfly-4hlyg)

**Comment by Jeyloh** [comment]
@Janaka-Steph I am the author of the article and I still haven't found a production-friendly solution. Loading/removing whole routes with timeouts seems too dangerous to add in production. 

I think the issue lies within the preload of a component. Even though the component is preloaded, it still hasn't been in the DOM, which is something required for the animation library to understand how to orchestrate the transition of a new route. That's why I tried to add/remove it in the DOM. 

I've also tried something like

```
const ReactLazyPreload = importStatement => {
  const Component = React.lazy(importStatement);
  Component.preload = importStatement;
  return Component;
};
const App = ReactLazyPreload(() =>
  import("./App")
);
App.preload()

```

but the same issue persists. It's not loaded in the DOM, therefore the first animation breaks. 

I'm struggling with framer-motion, but I don't think the animation library is to blame. Rather, it's the way lazy-loading works. I'll keep looking for a workaround, maybe within an alternative lazy-loading approach. If you guys find anything I would appreciate a notice :)

**Comment by PodaruDragos** [comment]
preloading kinda breaks the point of lazy-loading all together.

is this library not maintained anymore ? 
i did post this issues a lot of time ago tho..

**Comment by Jeyloh** [comment]
@PodaruDragos not really, if you're able to make a system which preloads every page which you'll use next. E.g.: A chain of components only available for certain users can be preloaded on login if that feature is available. If not, the user will never spend any mobile data on it.

**Comment by Jeyloh** [comment]
@Janaka-Steph I think I solved it!

By using this library: https://github.com/ianschmitz/react-lazy-with-preload you can preload components which will use the React.createElement and make the animation library get everything it needs to render the transition. This differs from React.lazy which only loads the JS, but doesn't invoke createElement.

For me, it works fine. I have it implemented in a custom hook like this:

```javascript 
if (user.authenticated) {
        authenticatedRoutes.forEach(route => {
            if (route.component.preload) route.component.preload();
        });
}
```

This runs in a useEffect, which depends on the user.authenticated, so as soon as it changes, all authenticatedRoutes are preloaded. Ergo, all my routes animates, but not until the user log in.

**Comment by Janaka-Steph** [comment]
@Jeyloh Oh good job! I gave up using React Lazy on my app but I will give this lib a try. 
Thank you!

**Comment by PsiKai** [comment]
Just like @Jeyloh I too used the `react-lazy-with-preload` library.  But I noticed when using it in `useEffect` my chain of requests was increasing some performance metrics like time to interactive.  Not much of a difference from real user perspective, but it was dinging my performance score for tools like Lighthouse.

To get around this, I preload a component when a user hovers over the nav-link for that component, like this:
```
const About = lazy(() => import('./pages/About'))

const preloadComponent = (component) => {
    component.preload()
  }

<li onMouseOver={() => preload(About)}>
     <Link to="/about">About</Link>
</li>
```

This works great for a desktop experience, and the chunk is loaded quickly enough to ensure that the page transition still fires.

 But to solve the issue of the mobile experience, where users can't really fire the `onMouseOver` effect with any efficacy, I instead decided to load all chunks when the user clicks the hamburger menu icon to see the nav-links.  I call a function `onClick` to loop through all the routes which I put into an array and preload each one:
```
const routes = [
    Main, About, Contact,
    Gallery, Cart, Login, Edit  
] 

const openSidenav = () => {
    routes.forEach(route => preload(route))
}

<div className="menu__burger-icon" onClick={openSidenav}>
</div>
```

My initial thought to this approach is that it would negate the purpose of lazy loading components since I still load them all at once, but after inspecting the network requests, I noticed that all the routes' chunks were downloading concurrently with a total time of ~10ms.  Each chunk is also very minimal in size (<1kb each), and this does not load the more expensive resources like images until the component is rendered.  Thusly, it preserves my performance increase without sacrificing my page transitions.
